man/vcpustat.8
.\" Copyright (C) 2020 International Business Machines.
.\"
.TH VCPUSTAT 8 "Mar 2020" Linux "Linux on Power Service Tools"
.SH NAME
vcpustat \- Report logical partition (LPAR) dispatch statistics.
.SH SYNOPSIS
.B /usr/sbin/vcpustat [ options ] [ <interval> [ count ] ]
.SH DESCRIPTION

The \fIvcpustat\fR command provides a report of logical processor (vcpu) dispatch statistics. This command provides a display of the logical processor dispatch dispersions, relative to the previous dispatch, as well as relative to the logical processor's home node as indicated by the Hypervisor.

The \fIvcpustat\fR command with no options will generate a single report containing raw dispatch counts since the last time statistics were enabled. This is useful if dispatch statistics have been previously enabled persistently using the -e option.

The following information is displayed in the report for each logical processor:
.TP
.B dispatch dispersions:
Represents logical processor dispatch dispersions, relative to the previous dispatch and includes the below columns:
.RS
.TP
total
Indicates the total number of dispatches of this logical processor. This is only displayed if numeric or raw values are being shown.
.TP
core
Indicates the number of times this logical processor was dispatched within the same core.
.TP
chip
Indicates the number of times this logical processor was dispatched in a different core, but within the same chip.
.TP
socket
Indicates the number of times this logical processor was dispatched in a different chip, but within the same socket. This is only seen on Multi Chip Modules.
.TP
cec
Indicates the number of times this logical processor was dispatched outside the socket.
.RE
.TP
.B numa dispersions:
Represents logical processor dispatch dispersions, relative to the logical processor's home node and includes the below columns:
.RS
.TP
home
Indicates the number of times this logical processor was dispatched in its home node.
.TP
adj
Indicates the number of times this logical processor was dispatched in the adjacent node.
.TP
far
Indicates the number of times this logical processor was dispatched in a further node.
.RE

.SH OPTIONS
.TP
\fB\-e, --enable\fR
Enable gathering dispatch statistics persistently, and not only when this command is active.
.TP
\fB\-d, --disable\fR
Disable gathering dispatch statistics.
.TP
\fB\-n, --numeric\fR
Display the values in numbers, rather than percentages.
.TP
\fB\-r, --raw\fR
Display the raw values, rather than the delta.
.TP
\fB\-h, --help\fR
Display the usage of vcpustat.
.TP
\fB\-V, --version\fR
Display the vcpustat version information.
.TP
interval
The
.B interval
parameter specifies the amount of time between each report.
.TP
count
The
.B count
parameter specifies how many reports will be displayed.

.SH NOTES
The preferred invocation of
.B vcpustat
is by specifying
.B interval
and an optional
.B count
to monitor logical processor dispatch dispersions over a desired interval of time. This enables in-kernel gathering of logical processor dispatch dispersions and provides a view of the same at the desired monitoring interval. With this form of invocation, the statistics are only gathered when the program is active resulting in minimal kernel overhead.

Long-term statistics can be obtained by enabling gathering of statistics in-kernel through the use of
.B -e
option, followed by invocations of
.B vcpustat
without any options, or using the
.B -r
option to display raw values, rather than the difference in each monitoring interval. This should only be used when long-term statistics are desired and results in additional kernel load. It is imperative that the statistics be disabled using
.B -d
option once monitoring is complete.

.SH FILES
.TP
.B /proc/powerpc/vcpudispatch_stats

.SH AUTHOR
Written by Naveen N. Rao <naveen.n.rao@linux.vnet.ibm.com>.


Changelog

powerpc-utils-1.3.6
=====================================================================

commit 1d4147370a148b3a2ebadcf02abd4a6c9a763e6d
Author: Sourabh Jain <sourabhjain@linux.ibm.com>
Date:   Sat Apr 27 10:18:32 2019 +0530

    lparstat: add an option to print the lparstat report similar to legacy lparstat tool
    
    This patch adds a new command line option (-l, --legacy) to allow
    users to print the report similar to legacy lprastat tool without
    affecting existing report format.
    
    Changes done to make the current report format similar to legacy
    report format are listed below:
    
    1) The memory attribute will be printed in MB instead of KB.
    2) cpus attribute will be referred as psize in legacy report format
       and only printed in shared lpar environment.
    3) On legacy lparstat tool the lcpu attribute is a combination of
       two attributes, smt and number of online processor. Formula used
       to calculate the lcpu is given below.
                  lcpu = smt * number of online processors
    
    Signed-off-by: Sourabh Jain <sourabhjain@linux.ibm.com>
    Reviewed-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

commit dd6da6b329bac2743d1c80b5556b494c923c11ad
Author: Michael Bringmann <mwb@linux.vnet.ibm.com>
Date:   Mon Apr 29 11:25:18 2019 -0500

    scripts: Improve handling of errors from subsidiary scripts
    
    This patch is to fix the handling of errors by 'lsdevinfo' from
    invocations of 'ofpathname' that also encounter errors.  Error
    messages from 'ofpathname' were being sent through 'stdout', and
    interpreted as valid data by a number of script instructions in
    'lsdevinfo'.  Following is an example of such a failure when the
    error results were forwarded to an associated HMC.
    
        2019-02-19 22:05:38.212 [TID-5f68--72195] VIOS_CACHE: CacheVios.<-> saveCacheData 3*9080-M9S*130A068 cmd=lsdevinfo -c -q physloc=U78CD.001.FZH0128-P1-C1-T3-S4 -F status
        2019-02-19 22:05:38.212 [TID-5f68--72195] RMC_VIOS: PIRmcViosClient.stdErr=/bin/ls: cannot access /sys/class/net/ofpathname:: No such file or directory
        /bin/ls: cannot access Could: No such file or directory
        /bin/ls: cannot access not: No such file or directory
        /bin/ls: cannot access retrieve: No such file or directory
        /bin/ls: cannot access logical: No such file or directory
        /bin/ls: cannot access device: No such file or directory
        /bin/ls: cannot access name: No such file or directory
        /bin/ls: cannot access for: No such file or directory
        /bin/ls: cannot access Open: No such file or directory
        /bin/ls: cannot access Firmware: No such file or directory
        /bin/ls: cannot access path: No such file or directory
        /bin/ls: cannot access "/pci@800000029008004/ethernet@0"./device/driver: No such file or directory
    
    This patch makes the following changes:
    
    * Redirects the error messages from 'ofpathname' to stderr
    * Adds more checks to 'lsdevinfo' to filter out stderr text
      from input to various parsing commands.
    
    Signed-off-by: Michael Bringmann <mwb@linux.vnet.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

commit 8b7978b114e5df89218daa9b4b48cc0e918ba917
Author: Sourabh Jain <sourabhjain@linux.ibm.com>
Date:   Mon Feb 4 16:15:14 2019 +0530

    lparstat: introduce the help command line option to print lparstat usage
    
    This patch adds a function to print the usage of lparstat. Also,
    a new element is added to structure option to handle long command
    line argument for help.
    
    Signed-off-by: Sourabh Jain <sourabhjain@linux.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

commit bc928b06605371b72c7fe8ec7e81886114a9e9d4
Author: Sourabh Jain <sourabhjain@linux.ibm.com>
Date:   Thu Jan 31 14:25:21 2019 +0530

    lparstat: restrict the physc and entc attribute values to two decimal places
    
    Print format of physical processors consumed (physc) and percentage of
    the entitled capacity consumed (entc) attributes has changed from six
    decimal places to two.
    
    The least possible value allowed for processor entitlement to a LPAR
    is 0.05 so printing physc and entc values with such a high precision
    is incorrect.
    
    Signed-off-by: Sourabh Jain <sourabhjain@linux.ibm.com>
    Reviewed-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

commit b5fea40ac98d1c9a685d98c011e88180c31ca0be
Author: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Date:   Wed Feb 20 15:44:17 2019 +0530

    man/update_flash: Add details to extract rpm format image
    
    Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

commit f567bdb8b2a3effdb4af06e2cf935b0350f2f881
Author: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Date:   Wed Feb 20 15:44:09 2019 +0530

    man/update_flash: Fix man page
    
    Add heading to NOTES section.
    
    Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>

powerpc-utils-1.3.6
=====================================================================

commit 01c9cbcf4face809cb6953bc3cd2a30cd075ed16
Author: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>
Date:   Thu Dec 6 17:30:38 2018 -0600

    lparstat: correct calculation of physc to use tbr
    
    Currently phsyc is calculated using the sampled PURR values from lparcfg and
    a derived timebase that lparstat itself calculates. However, both the Timer
    Facilities section in Book III of the ISA and PAPR outline that this resource
    calcuation is done by computing the ratio of the PURR delta and the Timebase
    Register delta for the sampled period.
    
    I've sent a patch upstream to the kernel list to also export the Timebase

src/vcpustat.c
/**
 * @file vcpustat.c
 * @brief vcpustat command
 *
 * Copyright (c) 2020 International Business Machines
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 * @author Naveen N. Rao <naveen.n.rao@linux.vnet.ibm.com>
 * 	   Adapted from 'lparstat.c'
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <signal.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/time.h>
#include "pseries_platform.h"

#define VCPUSTAT_FILE	"/proc/powerpc/vcpudispatch_stats"
#define NR_CPUS 4096

struct vcpudispatch_stat {
	unsigned long idx;
	int total;
	int same_cpu;
	int same_chip;
	int same_package;
	int diff_package;
	int home_numa_node;
	int next_numa_node;
	int far_numa_node;
};

unsigned long idx;
int retain_stats, numeric_stats, raw_stats, stats_off, intr;

int read_stats(struct vcpudispatch_stat stats[])
{
	struct vcpudispatch_stat stat;
	int cpu, rc = -1;
	char buf[144];
	FILE *f;

	f = fopen(VCPUSTAT_FILE, "r");
	if (!f) {
		fprintf(stderr, "Could not open %s\n", VCPUSTAT_FILE);
		return -1;
	}

	if ((fgets(buf, 144, f)) == NULL) {
		fprintf(stderr, "Could not read %s\n", VCPUSTAT_FILE);
		goto out;
	}

	if (!strncmp(buf, "off", 3)) {
		stats_off = 1;
		rc = 0; /* not an error */
		goto out;
	} else
		stats_off = 0;

	idx++;

	do {
		rc = sscanf(buf, "cpu%d %d %d %d %d %d %d %d %d", &cpu,
				  &stat.total, &stat.same_cpu, &stat.same_chip,
				  &stat.same_package, &stat.diff_package,
				  &stat.home_numa_node, &stat.next_numa_node,
				  &stat.far_numa_node);
		if (rc != 9) {
			fprintf(stderr, "Error parsing %s\n", VCPUSTAT_FILE);
			goto out;
		}
		if (cpu < 0 || cpu >= NR_CPUS) {
			fprintf(stderr, "Cpu (%d) out of range\n", cpu);
			goto out;
		}
		stat.idx = idx;
		stats[cpu] = stat;
	} while (!intr && (fgets(buf, 144, f)) != NULL);

	rc = 0;

out:
	fclose(f);
	return rc;
}

void print_alltime_stats(struct vcpudispatch_stat stats[])
{
	char raw_header1[] = "%22s %43s | %32s\n";
	char raw_header2[] = "%-7s | %10s | %10s %10s %10s %10s | %10s %10s %10s\n";
	char raw_fmt[] = "cpu%-4d | %10d | %10d %10d %10d %10d | %10d %10d %10d\n";
	int i;

	printf(raw_header1, " ",
				"========== dispatch dispersions ==========",
				"======= numa dispersions =======");
	printf(raw_header2, "cpu", "total",
					"core", "chip", "socket", "cec",
					"home", "adj", "far");

	for (i = 0; i < NR_CPUS; i++) {
		if (!stats[i].idx)
			continue;

		printf(raw_fmt, i,
			stats[i].total, stats[i].same_cpu, stats[i].same_chip,
			stats[i].same_package, stats[i].diff_package,
			stats[i].home_numa_node, stats[i].next_numa_node,
			stats[i].far_numa_node);
	}

	printf("\n");
	fflush(stdout);
}

void print_stats(struct vcpudispatch_stat stats1[],
		 struct vcpudispatch_stat stats2[])
{
	char percent_header1[] = "%35s | %20s\n";
	char percent_header2[] = "%-7s %6s %6s %6s %6s | %6s %6s %6s\n";
	char percent_fmt[] = "cpu%-4d %6.2f %6.2f %6.2f %6.2f | %6.2f %6.2f %6.2f\n";
	char raw_header1[] = "%22s %43s | %32s\n";
	char raw_header2[] = "%-7s | %10s | %10s %10s %10s %10s | %10s %10s %10s\n";
	char raw_fmt[] = "cpu%-4d | %10d | %10d %10d %10d %10d | %10d %10d %10d\n";
	struct vcpudispatch_stat stat;
	int i;

	if (stats_off) {
		printf("off\n");
		return;
	}

	if (numeric_stats || raw_stats) {
		printf(raw_header1, " ",
				"========== dispatch dispersions ==========",
				"======= numa dispersions =======");
		printf(raw_header2, "cpu", "total",
					"core", "chip", "socket", "cec",
					"home", "adj", "far");
	} else {
		printf(percent_header1, "         == dispatch dispersions ==",
					"= numa dispersions =");
		printf(percent_header2, "cpu", "core", "chip", "socket", "cec",
					"home", "adj", "far");
	}

	for (i = 0; i < NR_CPUS; i++) {
		if (stats2[i].idx != idx || stats1[i].idx != (idx - 1))
			continue;

		if (!raw_stats) {
			stat.total = stats2[i].total - stats1[i].total;
			stat.same_cpu = stats2[i].same_cpu - stats1[i].same_cpu;
			stat.same_chip = stats2[i].same_chip - stats1[i].same_chip;
			stat.same_package = stats2[i].same_package - stats1[i].same_package;
			stat.diff_package = stats2[i].diff_package - stats1[i].diff_package;
			stat.home_numa_node = stats2[i].home_numa_node - stats1[i].home_numa_node;
			stat.next_numa_node = stats2[i].next_numa_node - stats1[i].next_numa_node;
			stat.far_numa_node = stats2[i].far_numa_node - stats1[i].far_numa_node;
		}

		if (numeric_stats)
			printf(raw_fmt, i,
				stat.total, stat.same_cpu, stat.same_chip,
				stat.same_package, stat.diff_package,
				stat.home_numa_node, stat.next_numa_node,
				stat.far_numa_node);
		else if (raw_stats)
			printf(raw_fmt, i,
				stats2[i].total, stats2[i].same_cpu, stats2[i].same_chip,
				stats2[i].same_package, stats2[i].diff_package,
				stats2[i].home_numa_node, stats2[i].next_numa_node,
				stats2[i].far_numa_node);
		else
			printf(percent_fmt, i,
				100 * (float)stat.same_cpu / stat.total,
				100 * (float)stat.same_chip / stat.total,
				100 * (float)stat.same_package / stat.total,
				100 * (float)stat.diff_package / stat.total,
				100 * (float)stat.home_numa_node / stat.total,
				100 * (float)stat.next_numa_node / stat.total,
				100 * (float)stat.far_numa_node / stat.total);
	}

	printf("\n");
	fflush(stdout);
}

void process_stats(int interval, int count)
{
	struct vcpudispatch_stat *stats1, *stats2, *stats_tmp;
	int rc, dec = count;

	stats1 = calloc(NR_CPUS, sizeof(struct vcpudispatch_stat));
	stats2 = calloc(NR_CPUS, sizeof(struct vcpudispatch_stat));

	if (!stats1 || !stats2) {
		fprintf(stderr, "Error allocating memory for stats\n");
		return;
	}

	rc = read_stats(stats1);
	if (rc)
		goto out;
	sleep(interval);

	while (!intr) {
		rc = read_stats(stats2);
		if (rc)
			goto out;

		print_stats(stats1, stats2);

		stats_tmp = stats2;
		stats2 = stats1;
		stats1 = stats_tmp;

		if (count) {
			dec--;
			if (!dec)
				break;
		}

		if (!intr)
			sleep(interval);
	}

out:
	free(stats1);
	free(stats2);
}

void display_raw_counts(void)
{
	struct vcpudispatch_stat *stats;
	int rc;

	stats = calloc(NR_CPUS, sizeof(struct vcpudispatch_stat));

	if (!stats) {
		fprintf(stderr, "Error allocating memory for stats\n");
		return;
	}

	rc = read_stats(stats);
	if (rc)
		goto out;

	if (stats_off) {
		fprintf(stderr,
			"Dispatch statistics are not enabled. Please specify an interval to monitor.\n");
		goto out;
	}

	print_alltime_stats(stats);

out:
	free(stats);
}

int init_stats(bool enable, bool user_requested)
{
	char buf[144];
	int rc = 0;
	FILE *f;

	f = fopen(VCPUSTAT_FILE, "r+");
	if (!f) {
		fprintf(stderr, "Could not open %s\n", VCPUSTAT_FILE);
		return -1;
	}

	if ((fgets(buf, 144, f)) == NULL) {
		fprintf(stderr, "Could not read %s\n", VCPUSTAT_FILE);
		rc = -1;
		goto out;
	}

	if (strncmp(buf, "off", 3))
		retain_stats = 1;

	if (user_requested) {
		if (enable && retain_stats) {
			fprintf(stderr, "Stats are already enabled!\n");
			goto out;
		} else if (!enable && !retain_stats) {
			fprintf(stderr, "Stats are already disabled!\n");
			goto out;
		}
	}

	/* Enable stats */
	if (fprintf(f, "%d", enable) <= 0) {
		fprintf(stderr, "Couldn't setup stats: error writing to %s\n",
				VCPUSTAT_FILE);
		rc = -1;
	}

out:
	fclose(f);
	return rc;
}

void disable_stats(void)
{
	FILE *f;

	if (retain_stats)
		return;

	f = fopen(VCPUSTAT_FILE, "w");
	if (!f) {
		fprintf(stderr, "Couldn't disable stats: error opening %s\n",
			VCPUSTAT_FILE);
		return;
	}

	if (fprintf(f, "0") <= 0)
		fprintf(stderr, "Couldn't disable stats: error writing to %s\n",
			VCPUSTAT_FILE);

	fclose(f);
}

static void sighandler(int signum)
{
	intr = 1;
}

static void usage(void)
{
	printf("Usage:  vcpustat [ options ] [ <interval> [ count ] ]\n\n"
	       "options:\n"
	       "\t-e, --enable          Enable gathering statistics.\n"
	       "\t-d, --disable         Disable gathering statistics.\n"
	       "\t-n, --numeric         Display the statistics in numbers, rather than percentage.\n"
	       "\t-r, --raw             Display the raw counts, rather than the difference in an interval.\n"
	       "\t-h, --help            Show this message and exit.\n"
	       "\t-V, --version         Display vcpustat version information.\n"
	       "\tinterval              The interval parameter specifies the amount of time between each report.\n"
	       "\tcount                 The count parameter specifies how many reports will be displayed.\n");
}

static struct option long_opts[] = {
	{"version",	no_argument,		NULL,	'V'},
	{"help",	no_argument,		NULL,	'h'},
	{"enable",	no_argument,		NULL,	'e'},
	{"disable",	no_argument,		NULL,	'd'},
	{"numeric",	no_argument,		NULL,	'n'},
	{"raw",		no_argument,		NULL,	'r'},
	{0, 0, 0, 0},
};

int main(int argc, char *argv[])
{
	bool enable_only = false, disable_only = false;
	int platform = get_platform();
	int interval = 0, count = 0;
	struct sigaction sa;
	int c, opt_idx = 0;

	if (platform != PLATFORM_PSERIES_LPAR ||
			access(VCPUSTAT_FILE, F_OK) == -1) {
		if (platform == PLATFORM_PSERIES_LPAR)
			fprintf(stderr, "%s is not supported on this LPAR\n",
							argv[0]);
		else
			fprintf(stderr, "%s is not supported on the %s platform\n",
							argv[0], platform_name);
		exit(1);
	}

	while ((c = getopt_long(argc, argv, "Vhnred",
				long_opts, &opt_idx)) != -1) {
		switch (c) {
		case 'V':
			printf("vcpustat - %s\n", VERSION);
			return 0;
		case 'h':
			usage();
			return 0;
		case '?':
			usage();
			return 1;
		case 'e':
			enable_only = true;
			break;
		case 'd':
			disable_only = true;
			break;
		case 'n':
			numeric_stats = 1;
			break;
		case 'r':
			raw_stats = 1;
			break;
		default:
			break;
		}
	}

	/* see if there is an interval specified */
	if (optind < argc)
		interval = atoi(argv[optind++]);

	/* check for count specified */
	if (optind < argc)
		count = atoi(argv[optind++]);

	if (interval < 0 || count < 0) {
		fprintf(stderr, "Invalid interval/count specified\n");
		return -1;
	}

	if (enable_only && disable_only) {
		fprintf(stderr, "Please select only one of -e and -d\n");
		return -1;
	}

	if ((enable_only || disable_only) && (raw_stats || numeric_stats || interval)) {
		fprintf(stderr, "-e|-d cannot be used with other options\n");
		return -1;
	}

	if (enable_only || disable_only)
		return init_stats(enable_only, true);

	if (!interval) {
		display_raw_counts();
		return 0;
	}

	memset(&sa, 0, sizeof(sa));
	sigemptyset(&sa.sa_mask);
	sa.sa_handler = sighandler;

	if (sigaction(SIGINT, &sa, NULL) == -1) {
		fprintf(stderr, "Unable to setup signal handler\n");
		return -1;
	}

	if (init_stats(true, false))
		return -1;

	process_stats(interval, count);

	disable_stats();

	return 0;
}
